feat(ast_tools): generate TS declaration files for deserializer and walk files

camc314 opened the pull request:
this fixes some ts-expect-errors in oxlint by generating declaration files.

We could also consider using oxc_isolated_declarations to generate these

overlookmotel reviewed (approved):
I keep meaning to convert `napi/parser` to TS, and bundle it with TSDown. Then we could make all the generated files actual TS, and get better type safety / check the correctness of our types. But I haven't got around to it.

In the meantime, this is a good solution.

camc314 committed:
feat(ast_tools): generate TS declaration files for deserializer and walk files (#16912)

this fixes some ts-expect-errors in oxlint by generating declaration files.

We could also consider using oxc_isolated_declarations to generate these

camc314 commented:
> keep meaning to convert napi/parser to TS, and bundle it with TSDown. Then we could make all the generated files actual TS, and get better type safety / check the correctness of our types. But I haven't got around to it.


I was going to do this, but i fear it may be a pain.

The problem is that the minifier doesn't support TS, and we run the generated code through the minifier before formatting + writing to disk. I think the minification we need to do is fairly trivial, so maybe we just have a manual implementation.

overlookmotel commented:
> The problem is that the minifier doesn't support TS, and we run the generated code through the minifier before formatting + writing to disk. I think the minification we need to do is fairly trivial, so maybe we just have a manual implementation.

Ah ha good point.

I suppose if we're using TSDown, we don't actually need to run code through minifier in `ast_tools`, because TSDown will minify later. It's nice for the generated files to be post-minifier, because then it's easier to check minifier is doing what it's supposed to, but that's not essential.
